cocotb/runner.py

```diff
@@ -0,0 +1,694 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+
+
+import subprocess
+import os
+import sys
+import tempfile
+import re
+import shutil
+import warnings
+from xml.etree import cElementTree as ET
+import threading
+import signal
+import cocotb.config
+import sysconfig
+
+warnings.warn(
+    "Python runners and associated APIs are an experimental feature and subject to change.", UserWarning, stacklevel=2
+)
+
+_magic_re = re.compile(r"([\\{}])")
+_space_re = re.compile(r"([\s])", re.ASCII)
+
+
+def as_tcl_value(value):
+    # add '\' before special characters and spaces
+    value = _magic_re.sub(r"\\\1", value)
+    value = value.replace("\n", r"\n")
+    value = _space_re.sub(r"\\\1", value)
+    if value[0] == '"':
+        value = "\\" + value
+
+    return value
+
+
+class Simulator(object):
+    def __init__(self, build_dir="sim_build", work_dir=None):
+
+        self.build_dir = os.path.abspath(build_dir)
+        self.work_dir = self.build_dir
+
+        if work_dir is not None:
+            absworkdir = os.path.abspath(work_dir)
+            if os.path.isdir(absworkdir):
+                self.work_dir = absworkdir
+
+        self.env = {}
+        for e in os.environ:
+            self.env[e] = os.environ[e]
+
+        # Catch SIGINT and SIGTERM
+        self.old_sigint_h = signal.getsignal(signal.SIGINT)
+        self.old_sigterm_h = signal.getsignal(signal.SIGTERM)
+
+        # works only if main thread
+        if threading.current_thread() is threading.main_thread():
+            signal.signal(signal.SIGINT, self.exit_gracefully)
+            signal.signal(signal.SIGTERM, self.exit_gracefully)
+
+        self.process = None
+
+    def set_env(self):
+
+        for e in os.environ:
+            self.env[e] = os.environ[e]
+
+        self.env["LIBPYTHON_LOC"] = cocotb._vendor.find_libpython.find_libpython()
+
+        self.env["PATH"] += os.pathsep + cocotb.config.libs_dir
+
+        self.env["PYTHONPATH"] = os.pathsep.join(sys.path)
+        for path in self.python_search:
```

Comment:
Where is this attribute coming from? I don't see it in the `__init__`?

Reply:
https://github.com/cocotb/cocotb/blob/d3b1ec8538a33f706fccced0152de063114d90fa/cocotb/runner.py#L130

Only needed for finding tests. Maybe there is a better way?